python/lib/core/dmod/core/context/base.py

```diff
@@ -0,0 +1,197 @@
+"""
+@TODO: Put a module wide description here
+"""
+from __future__ import annotations
+
+import logging
+import typing
+import inspect
+import uuid
+import abc
+
+from ..common import format_stack_trace
+from ..common.protocols import LoggerProtocol
+
+T = typing.TypeVar('T')
+"""Represents some consistent yet generic type"""
+
+
+def is_property(obj: object, member_name: str) -> bool:
+    """
+    Checks to see if a member of an object is a property
+
+    Args:
+        obj: The object to check
+        member_name: The member on the object to check
+
+    Returns:
+        True if the member with the given name on the given object is a property
+    """
+    if not hasattr(obj, member_name):
+        raise AttributeError(f"{obj} has no attribute '{member_name}'")
+
+    if isinstance(obj, type):
+        return isinstance(getattr(obj, member_name), property)
+
+    # Is descriptor: inspect.isdatadescriptor(dict(inspect.getmembers(obj.__class__))[member_name])
+    parent_reference = dict(inspect.getmembers(obj.__class__))[member_name]
+    return isinstance(parent_reference, property)
+
+
+@typing.runtime_checkable
+class ObjectCreatorProtocol(typing.Protocol):
+    """
+    Defines the bare minimum methods that will be used that may create objects
+    """
+    @abc.abstractmethod
+    def create_object(self, name: str, /, *args, **kwargs) -> T:
+        """
+        Create an object and store its reference
+
+        Args:
+            name: The name of the class to instantiate
+            *args: Positional arguments used to instantiate the object
+            **kwargs: Keyword arguments used to instantiate the object
+
+        Returns:
+            A proxy pointing at the instantiated object
+        """
+
+    @abc.abstractmethod
+    def __str__(self):
+        ...
+
+    @abc.abstractmethod
+    def __repr__(self):
+        ...
+
+
+class ObjectManagerScope(abc.ABC, ObjectCreatorProtocol):
+    """
+    Maintains references to objects that have been instantiated via an object manager within a specific scope
+    """
+    def __init__(self, name: str, logger: LoggerProtocol = None):
+        self.__name = name
+        self.__items: typing.List[typing.Any] = []
+        self.__scope_id: uuid.UUID = uuid.uuid1()
+        self.__on_close: typing.List[
+            typing.Union[
+                typing.Callable[[ObjectManagerScope], typing.Any],
+                typing.Callable[[], typing.Any]
+            ]
+        ] = []
+        # Record the 4th frame as to where this scope started
+        # 0 = The `format_stack_trace` function
+        # 1 = This constructor
+        # 2 = The scope subclass' constructor
+        # 3 = Where the object was instantiated
+        self.__started_at: str = format_stack_trace(3)
```

Comment:
Debugging and tracing purposes. Distributed work like this can make it _really_ had to find what's launching what, so this helps indicate which Scope is being used and where.